crypto/evp/m_sha3.c

```diff
@@ -66,6 +66,9 @@ static int sha3_update(EVP_MD_CTX *evp_ctx, const void *_inp, size_t len)
     size_t bsz = ctx->block_size;
     size_t num, rem;
 
+    if (len == 0)
+        return 1;
+
```

Comment:
This check is analogous to [md32_common.h, L135](https://github.com/openssl/openssl/blob/master/crypto/include/internal/md32_common.h#L135).

Reply:
Actually, looking at the implemenation of `sha3_update()`, I can't see any undefined behaviour even in the absence of this check, since `memcpy(dest, src, 0)` is well behaved and a no-op, provided dest and src are valid pointers. What's your opinion @dot-asm?

Reply:
I'm sad. Because it's still a **non**-problem. Indeed, what does language standard say about memcpy? "The memcpy function copies n characters from the object pointed to by s2 into the object pointed to by s1. If copying takes place between objects that overlap, the behavior is undefined." Full stop. This means that passing zero length is in fact a no-op(*). Moreover, what does sanitizer documentation say? "While violating nullability does **not** have undefined behavior, it is often unintentional, so UBSan offers to catch it." Again, "does **not**"! This is just sad. It's sad to become a hostage of somebody else's idea about how bad programmers are. I mean it's offered with rationale that people would appreciate it, right? But program would crash anyway, so what [real] value does it actually add? ... Anyway, this means that whole premise that it's about undefined behaviour is essentially wrong. It's *not* about undefined behavour, but about sanitizer being overzealous. And with this in mind wouldn't it be more appropriate to just pass empty string as argument to first test in curve448_internal_test.c?

(*) One might be able to argue that memcpy(NULL,NULL,0) would be undefined, because NULL overlaps NULL. Even if one can, it's not case here, only one argument is NULL.

Reply:
> I almost anticipated that you would dislike this check,

It's rather that I dislike things that are done for wrong reason. But of course, if memcpy(0,0,0) is legitimately(*) undefined, then among all suggested solutions, the one in this request is most appropriate. At least you can observe the check and potentially problematic call on same page [nor does it affect general-purpose proxy interface]. For completeness sake it's possible to disable it selectively with
```
#ifdef __clang__
__attribute__((no_sanitize("undefined")))
#endif
static int sha3_update(...
```
But I'd say it would be too much compiler specifics. I mean code preferably should have minimum possible compiler dependencies.

(*) It should be noted that standard can have bugs too ;-)